pkg/listwatch/listwatch.go

```diff
@@ -44,56 +50,144 @@ import (
+	}
+
 	if IsAllNamespaces(allowedNamespaces) {
+		if !listWatchAllowed {
+			err := fmt.Errorf("missing list/watch permissions on the 'namespaces' resource")
+			for _, r := range reasons {
+				err = fmt.Errorf("%w: %w", err, r)
+			}
+
+			return nil, false, err
+		}
+
+		// deniedNamespaces is only supported with allowedNamespaces = "all".
 		tweak := func(options *metav1.ListOptions) {
 			DenyTweak(options, "metadata.name", deniedNamespaces)
 		}
+		if metadataNameLabelSupported {
+			// Using a label selector is more efficient but requires Kubernetes 1.22 at least.
+			tweak = func(options *metav1.ListOptions) {
+				TweakByLabel(options, "kubernetes.io/metadata.name", ExcludeFilterType, deniedNamespaces)
+			}
+		}
 
-		return cache.NewFilteredListWatchFromClient(c, "namespaces", metav1.NamespaceAll, tweak)
+		level.Debug(l).Log("msg", "using privileged namespace lister/watcher")
+		return cache.NewFilteredListWatchFromClient(
+			corev1Client.RESTClient(),
+			"namespaces",
+			metav1.NamespaceAll,
+			tweak,
+		), true, nil
+	}
+
+	if listWatchAllowed && metadataNameLabelSupported {
+		level.Debug(l).Log("msg", "using privileged namespace lister/watcher")
+		return cache.NewFilteredListWatchFromClient(
+			corev1Client.RESTClient(),
+			"namespaces",
+			metav1.NamespaceAll,
+			func(options *metav1.ListOptions) {
+				TweakByLabel(options, "kubernetes.io/metadata.name", IncludeFilterType, allowedNamespaces)
+			},
+		), true, nil
+	}
+
+	// At this point, the operator has no list/watch permissions on the
+	// namespaces resource. Check if it has at least the get permission to
+	// emulate the list/watch operations.
+	attrs := make([]k8sutil.ResourceAttribute, 0, len(allowedNamespaces))
+	for n := range allowedNamespaces {
+		attrs = append(attrs, k8sutil.ResourceAttribute{
+			Verbs:    []string{"get"},
+			Resource: "namespaces",
+			Name:     n,
+		})
+	}
+
+	getAllowed, reasons, err := k8sutil.IsAllowed(
+		ctx,
+		ssarClient,
+		nil, // namespaces is a cluster-scoped resource.
+		attrs...,
+	)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Only log a warning to preserve backward compatibility.
+	if !getAllowed {
+		err := fmt.Errorf("missing permissions")
+		for _, r := range reasons {
+			err = fmt.Errorf("%w: %w", err, r)
+		}
+
+		level.Warn(l).Log("msg", "the operator lacks required permissions which may result in degraded functionalities", "err", err)
 	}
 
 	listFunc := func(options metav1.ListOptions) (runtime.Object, error) {
 		list := &v1.NamespaceList{}
 		for name := range allowedNamespaces {
-			result := &v1.Namespace{}
-			err := c.Get().
-				Resource("namespaces").
-				Name(name).
-				VersionedParams(&options, scheme.ParameterCodec).
-				Do(ctx).
-				Into(result)
-			if apierrors.IsNotFound(err) {
-				level.Info(l).Log("msg", "namespace not found", "namespace", name)
-				continue
-			}
+			result, err := corev1Client.Namespaces().Get(ctx, name, metav1.GetOptions{})
```

Comment:
Should we put this inside an if?

```go
if getAllowed {
  result, err := corev1Client.Namespaces().Get(ctx, name, metav1.GetOptions{})
  .
  .
  .
}
```

Reply:
I kept the existing behavior to minimize the changes. The SA might have get permissions on a subset of the configured namespaces but getAllowed will be false.